packages/playwright-core/src/server/dom.ts

```diff
@@ -760,21 +782,32 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
     const { state = 'visible' } = options;
     if (!['attached', 'detached', 'visible', 'hidden'].includes(state))
       throw new Error(`state: expected one of (attached|detached|visible|hidden)`);
-    const info = this._page.parseSelector(selector, options);
-    const task = waitForSelectorTask(info, state, false, this);
     const controller = new ProgressController(metadata, this);
     return controller.run(async progress => {
       progress.log(`waiting for selector "${selector}"${state === 'attached' ? '' : ' to be ' + state}`);
-      const context = await this._context.frame._context(info.world);
-      const injected = await context.injectedScript();
-      const pollHandler = new InjectedScriptPollHandler(progress, await task(injected));
-      const result = await pollHandler.finishHandle();
-      if (!result.asElement()) {
-        result.dispose();
-        return null;
+
+      while (progress.isRunning()) {
+        const { frame, info } = await this._frame.resolveFrameForSelector(progress, selector, options, this);
+        // If we end up in the same frame => use the scope again, live above was noop.
+        const task = waitForSelectorTask(info, state, false, frame === this._frame ? this : undefined);
+        try {
+          const context = await frame._context(info.world);
+          const injected = await context.injectedScript();
+          const pollHandler = new InjectedScriptPollHandler(progress, await task(injected));
+          const result = await pollHandler.finishHandle();
+          if (!result.asElement()) {
+            result.dispose();
+            return null;
+          }
+          const handle = result.asElement() as ElementHandle<Element>;
+          return handle._adoptTo(await frame._mainContext());
+        } catch (e) {
+          // Navigated while trying to adopt the node.
+          if (js.isJavaScriptErrorInEvaluate(e) || isSessionClosedError(e))
```

Comment:
I think this now retries when |this| is detached, and it should not.

Reply:
No change in behavior here, added a failing test.